### PR TITLE
Feature: Encrypted traffic metadata

### PR DESCRIPTION
**Pull request by spendletonliveaction**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/5372

Describe changes:

This request adds a compile time feature to enable the generation of metadata that can be used to implement encrypted traffic analysis detection as described here:  https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/sec_data_eta/configuration/xe-16-6/sec-data-encrypted-traffic-analytics-xe-16-6-book/sec-data-encrypted-traffic-analytics-xe-16-6-book_chapter_01.html

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -570,6 +570,13 @@
         AC_DEFINE([HAVE_DETECT_DISABLED], [1], [Detection is disabled])
     ])
 
+  # enable response time calculations
```
Comment (victorjulien) [inline configure.ac:573]: I would like to avoid a configure time option for what can easily be a runtime option. Configure time options are really only meant for features that add new dependencies.

File: doc/userguide/output/eve/eve-json-output.rst [hunk 2]
```diff
@@ -460,6 +460,9 @@ The ``community-id-seed`` option specifies a unsigned 16 bit value that
 is used a seed to the hash that is calculated for the ``community-id``
 output. This must be set to the same value on all tools that output this
 record.
+The ``encrypted-traffic-metadata`` option is boolean. If set to ``true`` it
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-output.rst:463]: Can you expand this to include what the feature actually is, e.g. with a link to the doc linked in the PR for example.

File: src/flow.c [hunk 3]
```diff
@@ -378,6 +378,89 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
     }
 }
 
+#if defined(ENABLE_ETM)
+#include <math.h>
```
Comment (victorjulien) [inline src/flow.c:382]: system includes should go into `suricata-common.h` and if they are new they should be checked in `configure.ac`.

File: src/flow.c [hunk 4]
```diff
@@ -378,6 +378,89 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
...
+        splt->bd_variance = 0.0;
+    } else {
+        double variance = sqrt(splt->bd_variance / (splt->bd_count - 1));
+        splt->bd_variance = (float)variance;
+    }
+
+    /* final calculation of entropy */
+    uint32_t i;
+    float entropy = 0.0;
+    for (i = 0; i < FLOW_SPLT_BD_SIZE; i++) {
```
Comment (victorjulien) [inline src/flow.c:402]: nit: `i` can be declared in the for loop

File: src/flow.c [hunk 5]
```diff
@@ -378,6 +378,89 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
...
+        f->splt.dapp_bytes = p->payload_len;
+    else
+        f->splt.sapp_bytes = p->payload_len;
+
+    if (f->flow_state == FLOW_STATE_ESTABLISHED) {
+
+        if (f->splt.splt_count < FLOW_SPLT_MAX_COUNT) {
+            /* Update sequence of packet length & time (SPLT) */
+            f->splt.seq[f->splt.splt_count].dir = FlowGetPacketDirection(f, p);
+            f->splt.seq[f->splt.splt_count].len =
```
Comment (victorjulien) [inline src/flow.c:433]: `MIN(FLOW_SPLT_MAX_LEN,p->payload_len)`

File: src/flow.h [hunk 6]
```diff
@@ -302,6 +302,40 @@ typedef struct FlowCnf_
...
+typedef struct PacketSequence_
+{
+    uint32_t dir: 1;
+    uint32_t len: 13;
+    uint32_t delta: 18;
+} PacketSequence;
+
+typedef struct FlowSPLT_
+{
+    uint64_t first_epoch_msec;
```
Comment (victorjulien) [inline src/flow.h:323]: this seems to be set to the first packet in the `FLOW_STATE_ESTABLISHED`, so this will be when TCP state is ESTABLISHED or in other IP protos when we see the first packet in the toclient direction. Since this is not the same as `Flow::startts`, it seems to make sense to have it as a separate field.

File: src/flow.h [hunk 7]
```diff
@@ -302,6 +302,40 @@ typedef struct FlowCnf_
...
+    uint64_t last_epoch_msec;
```
Comment (victorjulien) [inline src/flow.h:324]: This field most likely is always synced to `Flow::lastts`, so is it really necessary?

File: src/flow.h [hunk 8]
```diff
@@ -302,6 +302,40 @@ typedef struct FlowCnf_
     SC_ATOMIC_DECLARE(uint64_t, memcap);
 } FlowConfig;
 
+#if defined(ENABLE_ETM)
+
+#define FLOW_SPLT_MAX_COUNT       32
```
Comment (victorjulien) [inline src/flow.h:307]: general comment for the defines and structs: add some documentation of that the fields and values are

File: src/flow.h [hunk 9]
```diff
@@ -503,6 +537,9 @@ typedef struct Flow_
     uint32_t tosrcpktcnt;
     uint64_t todstbytecnt;
     uint64_t tosrcbytecnt;
+#if defined(ENABLE_ETM)
+    FlowSPLT splt;
```
Comment (victorjulien) [inline src/flow.h:541]: this is a large member, so it would be probably have to use the `FlowStorage` API when the feature is done as a runtime option.

File: src/output-json-flow.c [hunk 10]
```diff
@@ -213,6 +213,44 @@ void EveAddFlow(Flow *f, JsonBuilder *js)
...
 }
 
+
+#if defined(ENABLE_ETM)
+void EveAddEncryptedTrafficMetaData(const Flow *f, JsonBuilder *js)
+{
+    if ((f->splt.first_epoch_msec > 0) && (f->splt.splt_count > 0)) {
+        uint32_t i = 0;
+        FlowEncryptedTrafficFinalize(f);
+        JsonBuilder *js_splt = jb_new_object();
```
Comment (victorjulien) [inline src/output-json-flow.c:223]: error check missing

File: src/flow.c [hunk 11]
```diff
@@ -378,6 +378,89 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
     }
 }
 
+#if defined(ENABLE_ETM)
+#include <math.h>
+#include <float.h>
+
+inline void FlowEncryptedTrafficFinalize(const Flow *f)
```
Comment (victorjulien) [inline src/flow.c:385]: this function has a lot of `const`, but ultimately does change the structure which seems to be bit of an anti pattern

File: src/output-json-flow.c [hunk 12]
```diff
@@ -213,6 +213,44 @@ void EveAddFlow(Flow *f, JsonBuilder *js)
     jb_set_string(js, "start", timebuf1);
 }
 
+
+#if defined(ENABLE_ETM)
+void EveAddEncryptedTrafficMetaData(const Flow *f, JsonBuilder *js)
+{
+    if ((f->splt.first_epoch_msec > 0) && (f->splt.splt_count > 0)) {
+        uint32_t i = 0;
+        FlowEncryptedTrafficFinalize(f);
```
Comment (victorjulien) [inline src/output-json-flow.c:222]: what is the purpose of finalize here? Why is it updating the flow from the output path? What happens if we have multiple eve instances and this gets called multiple times for the same packet?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Hi Stephen, thanks for this work. It looks interesting.

I've done an initial review.

Additionally:
- git commits will need to explain their changes and the functionality more
- CLA signature will be required, I'm not aware of one or some other agreement with your organization
- we'll need a ticket in our redmine
- we'll need tests in our Suricata-Verify suite

Thanks!

**Comment by victorjulien** [comment]
Looks like cifuzz found an int handling issue:
```
flow.c:435:53: runtime error: signed integer overflow: 7594870390325015023 * 1000 cannot be represented in type 'long'
13581
    #0 0x613662 in FlowEncryptedTrafficUpdate /src/suricata/src/flow.c:435:53
13582
    #1 0x613662 in FlowHandlePacketUpdate /src/suricata/src/flow.c:581:5
13583
    #2 0x6200f2 in FlowUpdate /src/suricata/src/flow-worker.c:208:5
13584
    #3 0x6200f2 in FlowWorker /src/suricata/src/flow-worker.c:505:17
13585
    #4 0x5b908f in LLVMFuzzerTestOneInput /src/suricata/src/tests/fuzz/fuzz_predefpcap_aware.c:130:13
13586
    #5 0x542d03 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
13587
    #6 0x5424ea in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
13588
    #7 0x544354 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
13589
    #8 0x544589 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
13590
    #9 0x5340bf in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
13591
    #10 0x55cdd2 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
13592
    #11 0x7fa893858082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
13593
    #12 0x50d1bd in _start (build-out/fuzz_predefpcap_aware+0x50d1bd)
```
https://github.com/OISF/suricata/runs/6519648111?check_suite_focus=true

**Comment by spendletonliveaction** [comment]
Thank you for the feedback. I am going through the issues. I added the redmine ticket here: [ ](https://redmine.openinfosecfoundation.org/issues/5372)

File: configure.ac (hunk 1 above)
Comment (spendletonliveaction) [reply configure.ac:573]: OK, I made it a configure time option because I thought adding a runtime option would incur some overhead to check if the feature was enabled or not during the run of the flow engine. I can make it a runtime option instead though.

File: doc/userguide/output/eve/eve-json-output.rst (hunk 2 above)
Comment (spendletonliveaction) [reply doc/userguide/output/eve/eve-json-output.rst:463]: Added some more explanation of the feature. I did not add the link because I was not sure if that link would be persistent (I do not work for Cisco). I can add it if you think it is helpful though.

File: configure.ac (hunk 1 above)
Comment (spendletonliveaction) [reply configure.ac]: Fixed with recent change

File: src/flow.c (hunk 5 above)
Comment (spendletonliveaction) [reply src/flow.c:431]: Fixed with recent change

File: src/flow.c (hunk 4 above)
Comment (spendletonliveaction) [reply src/flow.c]: Fixed with recent change

File: configure.ac [hunk 13]
```diff
@@ -68,7 +68,6 @@
     AC_PROG_AWK
     AC_PROG_CC
     AC_PROG_CPP
-    AC_PROG_RANLIB
```
Comment (victorjulien) [inline configure.ac:71]: unrelated change? Needs a dedicated commit

File: doc/userguide/partials/options.rst [hunk 14]
```diff
@@ -300,3 +300,12 @@
 .. option:: --simulate-ips
 
    Simulate IPS mode when running in a non-IPS mode.
+
+.. option:: --enable-etm
```
Comment (victorjulien) [inline doc/userguide/partials/options.rst:304]: I think I would prefer limiting configurability to the yaml. @jasonish thoughts?

File: src/flow.c [hunk 15]
```diff
@@ -378,6 +383,118 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
...
+                splt->seq[splt->splt_count].delta =
+                        (delta_epoch_msec > FLOW_SPLT_MAX_MSEC) ? FLOW_SPLT_MAX_MSEC : delta_epoch_msec;
+                splt->last_epoch_msec = now_epoch_msec;
+            }
+            splt->splt_count++;
+
+            /* Update byte distribution */
+            uint32_t i;
+            float delta;
+            for (i = 0; i < p->payload_len; i++) {
```
Comment (victorjulien) [inline src/flow.c:486]: nit: declare `i` in the for statement

File: src/flow.c [hunk 16]
```diff
@@ -378,6 +383,118 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
...
+                splt->bd_count++;
+                splt->bd[p->payload[i]]++;
+
+                delta = ((float)p->payload[i] - splt->bd_mean);
```
Comment (victorjulien) [inline src/flow.c:490]: nit: declare `delta` here

File: src/suricata.c [hunk 17]
```diff
@@ -1720,6 +1725,10 @@ static TmEcode ParseCommandLine(int argc, char** argv, SCInstance *suri)
                 if (suri->strict_rule_parsing_string == NULL) {
                     FatalError(SC_ERR_MEM_ALLOC, "failed to duplicate 'strict' string");
                 }
+            }  else if (strcmp((long_opts[option_index]).name, "enable-etm") == 0) {
+                SCLogInfo("Enabling encrypted traffic metadata");
```
Comment (victorjulien) [inline src/suricata.c:1729]: `SCLogConfig` seems more appropriate

**Review by victorjulien — COMMENTED** [review]
A few additional comments

File: doc/userguide/partials/options.rst (hunk 14 above)
Comment (jasonish) [reply doc/userguide/partials/options.rst:304]: I agree.  Config file option.  It can then be `--set` if needed.

File: doc/userguide/partials/options.rst [hunk 18]
```diff
@@ -299,4 +299,4 @@
 
 .. option:: --simulate-ips
 
-   Simulate IPS mode when running in a non-IPS mode.
+   Simulate IPS mode when running in a non-IPS mode.
```
Comment (inashivb) [inline doc/userguide/partials/options.rst:302]: nit: can be removed

File: src/flow.c [hunk 19]
```diff
@@ -378,6 +381,116 @@ static inline void FlowUpdateEthernet(ThreadVars *tv, DecodeThreadVars *dtv,
...
+{
+    FlowSPLT *const splt = (FlowSPLT *const)FlowGetStorageById(f, GetFlowSPLTInfoID());
+    if (splt->bd_entropy || (splt->splt_count == 0))
+        return;
+
+    /* final calculation of bd variance */
+    if (splt->bd_count == 1) {
+        splt->bd_variance = 0.0;
+    } else {
+        double variance = sqrt(splt->bd_variance / (splt->bd_count - 1));
```
Comment (inashivb) [inline src/flow.c:417]: Q: Could `splt->bd_count` be 0?

File: src/output-json.c [hunk 20]
```diff
@@ -1212,7 +1212,6 @@ OutputInitResult OutputJsonInitCtx(ConfNode *conf)
                         cid_seed);
             }
         }
-
```
Comment (inashivb) [inline src/output-json.c:1215]: nit: can be removed

**Comment by spendletonliveaction** [comment]
Added suricata-verify test for this feature here: https://github.com/OISF/suricata-verify/pull/857

**Comment by victorjulien** [comment]
Closing due to inactivity. If you're interested in picking this back up, please open a new PR addressing the comments. Thanks!
